Google/google_module/google_drive_reports.py

```diff
@@ -0,0 +1,203 @@
+            while self.is_running:
+                start = time.time()
+
+                try:
+                    self.get_repots_evts()
+
+                except HTTPError | BaseHTTPError as ex:
+                    self.log_exception(ex, message="Failed to get next batch of events")
+                except Exception as ex:
+                    self.log_exception(ex, message="An unknown exception occurred")
+                    raise
+
+                # compute the duration of the last events fetching
+                duration = int(time.time() - start)
+                FORWARD_EVENTS_DURATION.labels(intake_key=self.configuration.intake_key).observe(duration)
+
+                # Compute the remaining sleeping time
+                delta_sleep = self.configuration.frequency - duration
+                # if greater than 0, sleep
+                if delta_sleep > 0:
+                    time.sleep(delta_sleep)
+
+        finally:
+            self.log(
+                message="Failed to forward events from Google Reports API",
+                level="info",
+            )
+
+    def get_build_object(self):
+        credentials = service_account.Credentials.from_service_account_file(
+            self.service_account_path, scopes=self.scopes
+        )
+        delegated_credentials = credentials.with_subject(self.configuration.admin_mail)
+
+        reports_service = build("admin", "reports_v1", credentials=delegated_credentials)
+
+        return reports_service
+
+    def get_activities(self):
+        reports_service = self.get_build_object()
+
+        self.log(message=f"Start requesting the Goole reports with credential object created", level="info")
+
+        activities = (
+            reports_service.activities()
+            .list(
+                userKey="all",
+                applicationName=self.applicatioName,
+                maxResults=self.pagination_limit,
+                startTime=self.from_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+            )
+            .execute()
+        )
+
+        return activities
+
+    def get_next_activities(self, next_key):
+        reports_service = self.get_build_object()
+
+        self.log(message=f"Start requesting the Goole reports with credential object created", level="info")
+
+        activities = (
+            reports_service.activities()
+            .list(
+                userKey="all",
+                applicationName=self.applicatioName,
+                maxResults=self.pagination_limit,
+                pageToken=next_key,
+                startTime=self.from_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+            )
+            .execute()
+        )
+
+        return activities
+
+    def get_repots_evts(self):
+        now = datetime.now(timezone.utc)
+
+        self.log(message=f"Creating a google credentials objects", level="info")
+
+        activities = self.get_activities()
+
+        items = activities.get("items", [])
+        next_key = activities.get("nextPageToken")
```

Comment:
If the connector gets restarted, does it loose its progress and starts from the first page ?

Reply:
Good question, normally we stock the recent date seen in the cache. so if the connector restart in the middle of loop we will loose our progress and we will restart from the beginning. And that's because recent date is stocked just after the loop.

Reply:
Yes that's what I'm thinking I will do. with sorting items before looping.

Reply:
@TOUFIKIzakarya Maybe they already are sorted, or they could be an option for that in the request :+1: